### PR TITLE
arch: extract cleanup rules into pure computeCleanup() module

### LINKED ISSUE #9: arch: extract cleanup rules into a pure, deep module
## Problem

`applyCleanupRules()` is an 80-line async function (lines 132-278) mixing 6 business rules with interleaved Chrome API calls (query tabs, query groups, remove tabs, send notifications). The real bugs hide in the orchestration:

- Three separate whitelist checks with subtle differences (lines 148-158, 197, 222, 243)
- Inconsistent idle detection via `tab.lastAccessed` fallback vs custom `tabActivityMap`
- Tests use `@ts-nocheck` with 1616 lines of global mock setup

The function is untestable in isolation because Chrome is hardcoded.

## Solution

Extract the 6 rules into a pure function:

```ts
type CleanupResult = { tabIdsToClose: number[]; warningTriggered: boolean };
function computeCleanup(settings: Settings, tabs: Tab[], tabGroups: Map<number, TabGroup>, now: number, activityMap: Map<number, number>): CleanupResult
```

Business logic becomes deeply testable with zero mocks. Chrome calls stay in a thin adapter layer around the pure function.

## Benefits

- **Locality**: All cleanup logic concentrated in one pure module
- **Leverage**: Small interface, complex behaviour behind it
- **Tests shrink**: From 1616 lines of mock setup to direct assertions on input/output

## Files

- `src/background/index.ts` (lines 132-278)

### PR DESCRIPTION
**Pull request by pararang**
Extract the 6 business rules from applyCleanupRules() into a pure computeCleanup() function in src/background/cleanup.ts. getDuplicateTabs() also moved to cleanup.ts. applyCleanupRules() is now a thin wrapper that collects Chrome data, calls the pure function, then applies results.

Closes #9

### CODE REVIEW

**Review by pararang — COMMENTED** [review]
## Code Review — REQUEST CHANGES 🔴

*(Can't formally request changes on own PR — marking here.)*

The core extraction of `computeCleanup()` into `src/background/cleanup.ts` is well-executed: the function is genuinely pure (no chrome.* runtime calls, no side effects, no mutable global state), all 6 cleanup rules are preserved with identical logic, edge cases are handled (undefined groupId, missing lastAccessed, empty tabs), and there are no type safety violations (no `as any`, `@ts-ignore`). The wiring in `index.ts` correctly collects Chrome data, delegates to the pure function, and applies results. All 242 tests pass.

However, there is one significant **blocking issue**:

**🔴 Missing dedicated unit tests for `computeCleanup()`** — The primary motivation for extracting a pure function is testability. Yet there is no `cleanup.test.ts` file. The function has zero direct unit tests. It is only tested indirectly through the integration tests in `index.test.ts` which require Chrome API mocks, are slower, and cannot easily test edge cases. A dedicated test file should exercise `computeCleanup()` directly with plain data objects, no mocks needed. This is especially critical because this is the core business logic.

Test cases to cover:
- Empty tabs array
- All 6 rules independently (whitelist, blacklist, idle, duplicates, max-tabs, group whitelist)
- Rule interactions (e.g., whitelisted tab + blacklisted domain → whitelist wins)
- Edge cases: maxTabs=0 (unlimited), idleTimeout=0, tabs with undefined groupId/url/lastAccessed
- shouldWarn boundary conditions

**Minor observations (non-blocking):**
1. `CleanupInput` uses `whitelistedGroupIds: Set<number>` (pre-resolved) and `getLastActivity` (injected callback) — better design than originally described, more testable.
2. Warning reset condition simplified from `allTabs.length < warningThreshold && warningActive` to just `warningActive`. Semantic change (appears to be a bug fix) — worth being intentional about.
3. PR scope larger than title suggests (also removes hexagonal layer).

---
*Reviewed by Hermes Agent*

File: src/background/cleanup.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import type { Settings } from '../shared/settings';
```
Comment (pararang) [inline src/background/cleanup.ts:1]: **Priority: Add `cleanup.test.ts`** — This pure function is the crown jewel of this PR but has zero direct unit tests. The whole point of extracting it is testability. Please add a test file that exercises `computeCleanup()` directly with plain data — no Chrome mocks needed.

**Review by pararang — COMMENTED** [review]
Inline comments on the cleanup extraction.

**Comment by pararang** [comment]
Addressed review feedback: `src/background/cleanup.test.ts` added with 33 dedicated unit tests covering all 6 rules independently, rule interactions (whitelist > blacklist, group whitelist > blacklist), and edge cases (maxTabs=0, idleTimeout=0, undefined url/lastAccessed, warning threshold boundaries). Commit 3e11971.

**Comment by pararang** [comment]
Another push to fix Node 22.x build: added missing required `chrome.tabs.Tab` fields (`frozen`, `selected`, `discarded`, `autoDiscardable`) to the test helper.

**Comment by pararang** [comment]
## Re-review — APPROVE ✅

The blocking issue from the previous review has been addressed: **`cleanup.test.ts` now exists** with 380 lines of direct unit tests. No Chrome mocks needed — exactly as intended.

### What's verified

**Pure function correctness:**
- `computeCleanup()` is genuinely pure — zero `chrome.*` runtime calls, no side effects, no mutable global state
- All 6 rules preserved with identical logic to the original `applyCleanupRules()`
- `chrome.tabs.Tab` usage is type-only (no runtime dependency on Chrome types at execution)

**Wiring in `index.ts`:**
- `applyCleanupRules()` is now a thin wrapper: collects Chrome data → calls `computeCleanup()` → applies results
- Tab group name→ID resolution happens before the pure call (correct)
- Warning state management and notification side effects stay in `index.ts` (correct)
- `index.ts` dropped from ~374 lines to 274 lines — real reduction

**Test coverage in `cleanup.test.ts` (26 test cases):**
- ✅ Empty/no-op (4 tests) — empty tabs, all active, all pinned, no id
- ✅ Rule 1: Tab group whitelist (3 tests) — whitelisted, non-whitelisted, groupId=-1
- ✅ Rule 2: Domain whitelist (2 tests) — whitelisted beats blacklist, non-matching
- ✅ Rule 3: Blacklist (3 tests) — match, no match, subdomain match
- ✅ Rule 4: Idle timeout (4 tests) — idle, recently active, custom getLastActivity, idleTimeout=0
- ✅ Rule 5: Duplicate tabs (4 tests) — keep recent, keep active, pinned, no URL
- ✅ Rule 6: Max tabs (4 tests) — eviction, maxTabs=0 unlimited, whitelist respected, active protected
- ✅ Rule interactions (3 tests) — whitelist beats blacklist, group whitelist beats blacklist, duplicate respects whitelist
- ✅ Warning threshold (3 tests) — warn at threshold, below threshold, maxTabs too small
- ✅ Edge cases (3 tests) — undefined url, undefined lastAccessed, tabCount

**All 273 tests pass** (273 passed, 2 skipped).

### Minor observations (non-blocking)

1. **Line 96 comment accuracy**: `// 3. Blacklist — close blacklisted domains (even if recently active)` — the comment says "even if recently active", but line 86 already skips `tab.active` tabs. Blacklisted tabs are only closed if *inactive*. The comment is slightly misleading — it should say "if inactive" not "even if recently active".

2. **`getDuplicateTabs` re-verifies whitelist (lines 109-120)**: Rule 5 calls `getDuplicateTabs()` which doesn't filter by whitelist, then re-filters the results. This is correct behavior (whitelisted duplicates should be kept) but the double-pass is slightly redundant. A future refactor could have `getDuplicateTabs` accept a filter predicate to avoid the extra O(n) scan.

3. **Line 104: idle check uses `lastActivity &&`**: This means tabs with `lastActivity === 0` (never tracked) are never closed by idle timeout. This is intentional — newly opened tabs without activity records shouldn't be immediately closed — but it's not documented.

---
*Reviewed by Hermes Agent*

File: src/background/cleanup.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    const domain = getDomain(tab.url);
+
+    // 1. Tab group whitelist — skip tabs in whitelisted groups
+    if (tab.groupId && whitelistedGroupIds.has(tab.groupId)) return;
+
+    // 2. Domain whitelist — skip whitelisted domains
+    if (whitelist.some((w) => domainMatches(domain, w))) return;
+
+    // 3. Blacklist — close blacklisted domains (even if recently active)
```
Comment (pararang) [inline src/background/cleanup.ts:96]: Comment says "close blacklisted domains (even if recently active)" but line 86 already skips `tab.active` tabs. Blacklisted tabs are only closed if **inactive**. Suggest: `// 3. Blacklist — close inactive tabs on blacklisted domains`

File: src/background/cleanup.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    if (blacklist.some((b) => domainMatches(domain, b))) {
+      tabsToClose.add(tab.id);
+      return;
+    }
+
+    // 4. Idle timeout — close if idle beyond configured timeout
+    const lastActivity = getLastActivity(tab);
+    if (lastActivity && now - lastActivity > idleTimeoutMs) {
```
Comment (pararang) [inline src/background/cleanup.ts:104]: The `lastActivity &&` guard means tabs with `lastActivity === 0` (never tracked) are never closed by idle timeout. This is correct behavior (don't close newly-opened tabs) but not documented — consider adding a brief comment explaining the guard.

**Review by pararang — COMMENTED** [review]
Inline observations on cleanup.ts
